modules/packages/Channel.chpl

```diff
@@ -0,0 +1,574 @@
+                var process$ : single bool;
+                var processing = new unmanaged Waiter(val, process$);
+
+                sendWaiters.enque(processing);
+                unlock();
+                var status = processing.suspend();
+                delete processing;
+                if status == false {
+                    throw new owned ChannelError("Trying to send on a closed channel");
+                }
+                return status;
+
+            }
+            else {
+                if !recvWaiters.isEmpty() {
+                    var receiver = recvWaiters.deque();
+                    receiver.val.deref() = val;
+
+                    receiver.release(true);
+                }
+                else {
+                    buffer[sendIdx] = val;
+
+                    sendIdx = (sendIdx + 1) % bufferSize;
+                    count += 1;
+                }
+
+                if !selected then unlock();
+                return true;
+            }
+        }
+
+        /*
+        This function is used to close a channel indicating that no more data
+        can be sent to it.
+
+        :throws ChannelError: If called on a closed channel.
+        */
+
+        proc close() throws {
+
+            lock();
+            if closed {
+                unlock();
+                throw new owned ChannelError("Trying to close a closed channel");
+            }
+            closed = true;
+            var queued = new WaiterQue(eltType);
+            while(!recvWaiters.isEmpty()) {
+                var receiver = recvWaiters.deque();
+                if(receiver.isSelect) {
+                    if receiver.isSelectDone.deref().compareAndSwap(-1, receiver.selectId) then 
+                        queued.enque(receiver);
+                }
+                else queued.enque(receiver);
+            }
+
+            while(!sendWaiters.isEmpty()) {
+                var sender = sendWaiters.deque();
+                if(sender.isSelect) {
+                    if sender.isSelectDone.deref().compareAndSwap(-1, sender.selectId) then
+                        queued.enque(sender);
+                }
+                else queued.enque(sender);
+            }
+            unlock();
+
+            while(!queued.isEmpty()) {
+                var waiter = queued.deque();
+                waiter.release(false);
+            }
+        }
+
+        /* Iterator to receive data from the channel until it is closed. */
+        iter these() {
+            while(true) {
+                var received : eltType;
+                var status = recv(received);
+                if !status then break;
+                yield received;
+            }
+        }
+    }
+
+    /* Error class for Channel */
+    pragma "no doc"
+    class ChannelError : Error {
+        var msg:string;
+
+        proc init(msg: string) {
+            this.msg = msg;
+        }
+
+        override proc message() {
+            return msg;
+        }
+    }
+
```

Comment:
At this point it would be good to put a block comment that explains a bit about the select statement & how it is implemented here. I'd like to request that you also include the contents of https://github.com/Yudhishthira1406/Chapel-Channel/blob/select-deq/SELECT.md in that comment so that we have the details here next to the implementation.